fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java

```diff
@@ -295,7 +297,8 @@ private PlanNode createQueryPlan(QueryStmt stmt, Analyzer analyzer, long default
             ((SortNode) root).setDefaultLimit(limit == -1);
             ((SortNode) root).setOffset(stmt.getOffset());
             if (useTopN) {
-                root.setLimit(limit != -1 ? limit : newDefaultOrderByLimit);
+                // if limit != -1, it means limit clause exists, we get the minimum.
+                root.setLimit(limit != -1 ? limit : Math.min(newDefaultOrderByLimit, sqlSelectLimit));
```

Comment:
sqlSelectLimit could small than 0, in this case, sqlSelectLimit should not be set to limit